src/python/pants/core/target_types.py

```diff
@@ -42,19 +57,114 @@
 from pants.util.docutil import bin_name
 from pants.util.frozendict import FrozenDict
 from pants.util.logging import LogLevel
+from pants.util.meta import frozen_after_init
 from pants.util.strutil import softwrap
 
 # -----------------------------------------------------------------------------------------------
-# `file` and`files` targets
+# Asset target helpers
 # -----------------------------------------------------------------------------------------------
 
 
-class FileSourceField(SingleSourceField):
+@dataclass(unsafe_hash=True)
+@frozen_after_init
+class HTTPSource:
+    url: str
+    len: int
+    sha256: str
+    # Defaults to last part of the URL path (E.g. `index.html`)
+    filename: str
+
+    def __init__(self, url: str, *, len: int, sha256: str, filename: str = ""):
```

Comment:
nit: `None` is more explicit for `filename`

Reply:
I did this because it makes the code simpler. Otherwise the type validation fails.

Moving the validation below the assignment means the URL being wrong type breaks file ame default (can't split an int or whatever else)

Reply:
I did this because it makes the code simpler. Otherwise the type validation fails.

Moving the validation below the assignment means the URL being wrong type breaks file ame default (can't split an int or whatever else)